// expo-jwt/dist/errors.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InvalidAudience = exports.InvalidIssuer = exports.InvalidSubject = exports.IssuedAtInvalid = exports.ImmatureSignature = exports.TokenExpired = exports.AlgorithmNotSupported = exports.AlgorithmMissing = exports.InvalidBody = exports.InvalidHeader = exports.SignatureInvalid = exports.InvalidStructure = undefined;

var _algorithms = require('./algorithms');

function _extendableBuiltin23(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _extendableBuiltin21(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _extendableBuiltin19(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _extendableBuiltin17(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _extendableBuiltin15(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _extendableBuiltin13(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _extendableBuiltin11(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _extendableBuiltin9(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _extendableBuiltin7(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _extendableBuiltin5(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _extendableBuiltin3(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _extendableBuiltin(cls) {
  function ExtendableBuiltin() {
    var instance = Reflect.construct(cls, Array.from(arguments));
    Object.setPrototypeOf(instance, Object.getPrototypeOf(this));
    return instance;
  }

  ExtendableBuiltin.prototype = Object.create(cls.prototype, {
    constructor: {
      value: cls,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });

  if (Object.setPrototypeOf) {
    Object.setPrototypeOf(ExtendableBuiltin, cls);
  } else {
    ExtendableBuiltin.__proto__ = cls;
  }

  return ExtendableBuiltin;
}

var InvalidStructure = exports.InvalidStructure = function (_extendableBuiltin2) {
  _inherits(InvalidStructure, _extendableBuiltin2);

  function InvalidStructure() {
    _classCallCheck(this, InvalidStructure);

    var _this = _possibleConstructorReturn(this, (InvalidStructure.__proto__ || Object.getPrototypeOf(InvalidStructure)).call(this, arguments));

    _this.message = 'Token must be in the format: header.body.signature';
    return _this;
  }

  return InvalidStructure;
}(_extendableBuiltin(Error));

var SignatureInvalid = exports.SignatureInvalid = function (_extendableBuiltin4) {
  _inherits(SignatureInvalid, _extendableBuiltin4);

  function SignatureInvalid() {
    _classCallCheck(this, SignatureInvalid);

    var _this2 = _possibleConstructorReturn(this, (SignatureInvalid.__proto__ || Object.getPrototypeOf(SignatureInvalid)).call(this, arguments));

    _this2.message = 'Invalid token signature';
    return _this2;
  }

  return SignatureInvalid;
}(_extendableBuiltin3(Error));

var InvalidHeader = exports.InvalidHeader = function (_extendableBuiltin6) {
  _inherits(InvalidHeader, _extendableBuiltin6);

  function InvalidHeader() {
    _classCallCheck(this, InvalidHeader);

    var _this3 = _possibleConstructorReturn(this, (InvalidHeader.__proto__ || Object.getPrototypeOf(InvalidHeader)).call(this, arguments));

    _this3.message = 'Invalid token header format';
    return _this3;
  }

  return InvalidHeader;
}(_extendableBuiltin5(Error));

var InvalidBody = exports.InvalidBody = function (_extendableBuiltin8) {
  _inherits(InvalidBody, _extendableBuiltin8);

  function InvalidBody() {
    _classCallCheck(this, InvalidBody);

    var _this4 = _possibleConstructorReturn(this, (InvalidBody.__proto__ || Object.getPrototypeOf(InvalidBody)).call(this, arguments));

    _this4.message = 'Invalid token body format';
    return _this4;
  }

  return InvalidBody;
}(_extendableBuiltin7(Error));

var AlgorithmMissing = exports.AlgorithmMissing = function (_extendableBuiltin10) {
  _inherits(AlgorithmMissing, _extendableBuiltin10);

  function AlgorithmMissing() {
    _classCallCheck(this, AlgorithmMissing);

    var _this5 = _possibleConstructorReturn(this, (AlgorithmMissing.__proto__ || Object.getPrototypeOf(AlgorithmMissing)).call(this, arguments));

    _this5.message = 'Algorithm specified in header is missing';
    return _this5;
  }

  return AlgorithmMissing;
}(_extendableBuiltin9(Error));

var AlgorithmNotSupported = exports.AlgorithmNotSupported = function (_extendableBuiltin12) {
  _inherits(AlgorithmNotSupported, _extendableBuiltin12);

  function AlgorithmNotSupported() {
    _classCallCheck(this, AlgorithmNotSupported);

    var _this6 = _possibleConstructorReturn(this, (AlgorithmNotSupported.__proto__ || Object.getPrototypeOf(AlgorithmNotSupported)).call(this, arguments));

    var supported = _algorithms.supportedAlgorithms.join(', ');
    _this6.message = 'Algorithm not supported. Supported algorithms are: ' + supported;
    return _this6;
  }

  return AlgorithmNotSupported;
}(_extendableBuiltin11(Error));

var TokenExpired = exports.TokenExpired = function (_extendableBuiltin14) {
  _inherits(TokenExpired, _extendableBuiltin14);

  function TokenExpired() {
    _classCallCheck(this, TokenExpired);

    var _this7 = _possibleConstructorReturn(this, (TokenExpired.__proto__ || Object.getPrototypeOf(TokenExpired)).call(this, arguments));

    _this7.message = 'Token has expired';
    return _this7;
  }

  return TokenExpired;
}(_extendableBuiltin13(Error));

var ImmatureSignature = exports.ImmatureSignature = function (_extendableBuiltin16) {
  _inherits(ImmatureSignature, _extendableBuiltin16);

  function ImmatureSignature() {
    _classCallCheck(this, ImmatureSignature);

    var _this8 = _possibleConstructorReturn(this, (ImmatureSignature.__proto__ || Object.getPrototypeOf(ImmatureSignature)).call(this, arguments));

    _this8.message = 'Token used before becoming valid';
    return _this8;
  }

  return ImmatureSignature;
}(_extendableBuiltin15(Error));

var IssuedAtInvalid = exports.IssuedAtInvalid = function (_extendableBuiltin18) {
  _inherits(IssuedAtInvalid, _extendableBuiltin18);

  function IssuedAtInvalid() {
    _classCallCheck(this, IssuedAtInvalid);

    var _this9 = _possibleConstructorReturn(this, (IssuedAtInvalid.__proto__ || Object.getPrototypeOf(IssuedAtInvalid)).call(this, arguments));

    _this9.message = 'Issued At claim is invalid';
    return _this9;
  }

  return IssuedAtInvalid;
}(_extendableBuiltin17(Error));

var InvalidSubject = exports.InvalidSubject = function (_extendableBuiltin20) {
  _inherits(InvalidSubject, _extendableBuiltin20);

  function InvalidSubject() {
    _classCallCheck(this, InvalidSubject);

    var _this10 = _possibleConstructorReturn(this, (InvalidSubject.__proto__ || Object.getPrototypeOf(InvalidSubject)).call(this, arguments));

    _this10.message = 'Subject is invalid';
    return _this10;
  }

  return InvalidSubject;
}(_extendableBuiltin19(Error));

var InvalidIssuer = exports.InvalidIssuer = function (_extendableBuiltin22) {
  _inherits(InvalidIssuer, _extendableBuiltin22);

  function InvalidIssuer() {
    _classCallCheck(this, InvalidIssuer);

    var _this11 = _possibleConstructorReturn(this, (InvalidIssuer.__proto__ || Object.getPrototypeOf(InvalidIssuer)).call(this, arguments));

    _this11.message = 'Issuer is invalid';
    return _this11;
  }

  return InvalidIssuer;
}(_extendableBuiltin21(Error));

var InvalidAudience = exports.InvalidAudience = function (_extendableBuiltin24) {
  _inherits(InvalidAudience, _extendableBuiltin24);

  function InvalidAudience() {
    _classCallCheck(this, InvalidAudience);

    var _this12 = _possibleConstructorReturn(this, (InvalidAudience.__proto__ || Object.getPrototypeOf(InvalidAudience)).call(this, arguments));

    _this12.message = 'Audience is invalid';
    return _this12;
  }

  return InvalidAudience;
}(_extendableBuiltin23(Error));
